experimental/dds/tree/src/interface-prototype/Examples.ts

```diff
@@ -0,0 +1,141 @@
+	Command,
+	CommandContext,
+	CommandId,
+	CommandRegistry,
+	PrefetchFilter,
+	root,
+	SharedTree,
+	TreeNodeViewReadonly,
+} from './Checkout';
+import { Place, TreeNode } from './MutableAnchors';
+
+//////////////// Command examples //////////////
+
+// Inserts a node with the specified Definition and identifier at the specified Place, and return it.
+export const insertExample = {
+	id: 'f73f004c-3b3e-42fe-b7c9-2e5e8793ca45' as CommandId,
+	run: (
+		context: CommandContext,
+		{ definition: def, identifier: id }: { definition: StableId; identifier: StableId },
+		{ place }: { place: Place }
+	): TreeNode => {
+		const definition = context.loadDefinition(def);
+		const identifier = context.loadNodeId(id);
+		const range = place.insert({ definition, identifier, traits: {} });
+		return range.start.adjacentNode(Side.After);
+	},
+};
+
+function newNodeId(): StableId {
+	return v4() as StableId;
+}
+
+export const doubleInsertExample = {
+	id: '08bac27d-632f-48bb-834a-90af8d67ca60' as CommandId,
+	run: (context: CommandContext, _: {}, { parent }: { parent: TreeNode }): TreeNode => {
+		const a = context.runCommand(
+			insertExample,
+			{ definition: context.stabilize(bar), identifier: newNodeId() },
+			{ place: parent.childrenFromTrait(testTrait).start }
+		);
+
+		return context.runCommand(
+			insertExample,
+			{ definition: context.stabilize(baz), identifier: newNodeId() },
+			{ place: a.childrenFromTrait(testTrait).start }
+		);
+	},
+};
+
+const filterAll = () => true;
+
+const fetchAll: PrefetchFilter = {
+	value: filterAll,
+	children: filterAll,
+	traitChunks: filterAll,
+};
+
+const commands: CommandRegistry = [insertExample, doubleInsertExample];
+
+// Some dummy schema related data.
+const foo: Definition = 'Foo' as Definition;
+const bar: Definition = 'Bar' as Definition;
+const baz: Definition = 'Bar' as Definition;
+
+const testTrait: TraitLabel = 'testTrait' as TraitLabel;
+
+export async function exampleApp(tree: SharedTree): Promise<void> {
+	// Perform a full checkout of all data.
+	const checkout = await tree.checkout(commands, fetchAll);
+
+	// Example app policy: this app just watches the tree, and adds bar(baz) subtree under any new `foo` node.
+	checkout.on('viewChange', (before, after) => {
+		const delta = before.delta(after);
+		for (const added of delta.added) {
+			const n: TreeNodeViewReadonly = checkout.contextualizeAnchor(anchorDataFromNodeId(added));
+			if (n.definition === foo) {
+				checkout.runCommand(doubleInsertExample, {}, { parent: n as TreeNodeData });
+			}
+		}
+	});
+
+	const treeRoot: TreeNodeViewReadonly = checkout.contextualizeAnchor(root);
+	console.log(treeRoot.queryJsonSnapshot.subtree);
+	await wait(10000);
+	console.log(treeRoot.queryJsonSnapshot.subtree);
+}
+
+function wait(ms) {
+	return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+// Note: Redo is the same as undoing an undo (at this level).
+export const undo: Command<{ editId: StableId }, {}, void> = {
```

Comment:
Not sure why we'd put this in the examples. Won't this be built into SharedTree?

Reply:
Modifications to the tree are done through commands. Undo modifies the tree so it is a command.

Like other commands, it has some higher level logic that could be replayed, and thus can benefit from the hierarchal nature of commands.

Making it a command allows for a simpler abstraction at this level, and I see no downside.

If undo is going to be a command, I figured it would make a reasonable example of one: I figured that would be better than simply omitting it from this prototype. Its particularly interesting because of its history access requirements and the desire to make it synchronous to be compatible with existing users of experimental/tree.

Reply:
Hm, I'm not sure an app *could* write undo, once we get through all the most complex scenarios. I don't think we should have this in the examples. Let's pick something else instead.